### PR TITLE
Implementation of Coreputation contract inheritance, storage and enums

### LINKED ISSUE #164: Implementation of Coreputation contract inheritance and storage, enums
1. `CoconsensusModule` will have a fixed constant `Coconsensus` address from the genesis.
2. Coconsensus.setup method is not needed since `Coconsensus` address is constant.
3. Define `ValidatorStatus` enum. ValidatorStatus enum values are `Undefined`, `Slashed`, `Staked`, `Deregistered`.
4. Define `ValidatorInfo` struct. Proposed state variables in the struct are below:
    - Validator status
    - Reputation value
5. Define the mapping of the validator address versus ValidatorInfo.
    ```solidity
    mapping(address => ValidatorInfo) public validators
    ```
6. The CoReputation contract should follow the proxy pattern for deployment.
7. `Coreputation` is `MasterCopyNonUpgradable`, `CoconsensusModule`

### PR DESCRIPTION
**Pull request by abhayks1**
PR covers below:

1. Define `ValidatorStatus` enum. ValidatorStatus enum values are `Undefined`, `Slashed`, `Staked`, `Deregistered`.
2. Define `ValidatorInfo` struct. Proposed state variables in the struct are below:
    - Validator status
    - Reputation value
3. Define the mapping of the validator address versus ValidatorInfo.
    ```solidity
    mapping(address => ValidatorInfo) public validators
    ```
4. The CoReputation contract should follow the proxy pattern for deployment.
5. `Coreputation` is `MasterCopyNonUpgradable`, `CoconsensusModule`.
6. Added copyright

**Note**: Added a TODO to Remove setup method after coConsensus storage variable is made
constant value in CoConsensusModule. setup method is kept for testing purpose. 

Fixes #164

### CODE REVIEW

File: contracts/reputation/Coreputation.sol [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        /** Undefined as null value */
+        Undefined,
+
+        /** Validator has been slashed and lost stake and earnings */
+        Slashed,
+
+        /** Validator has put up stake and participates in consensus */
+        Staked,
+
+        /** Validator has deregistered and no longer participates in consensus */
```
Comment (gulshanvasnani) [inline contracts/reputation/Coreputation.sol]:
```suggestion
        /** Validator has been deregistered and no longer participates in consensus */
```

File: contracts/reputation/Coreputation.sol [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import "../consensus/CoConsensusModule.sol";
+import "../proxies/MasterCopyNonUpgradable.sol";
+
+contract Coreputation is MasterCopyNonUpgradable, CoConsensusModule {
```
Comment (gulshanvasnani) [inline contracts/reputation/Coreputation.sol:26]: Please add documentation for the contract.

**Review by gulshanvasnani — COMMENTED** [review]
Nice 👍 
Couple of inline comment for documentation.

File: contracts/reputation/Coreputation.sol [hunk 3]
```diff
@@ -0,0 +1,70 @@
+pragma solidity >=0.5.0 <0.6.0;
+
+// Copyright 2019 OpenST Ltd.
```
Comment (gulshanvasnani) [inline contracts/reputation/Coreputation.sol]:
```suggestion
// Copyright 2020 OpenST Ltd.
```

Comment (abhayks1) [reply contracts/reputation/Coreputation.sol]: Done!

File: contracts/reputation/Coreputation.sol (hunk 2 above)
Comment (abhayks1) [reply contracts/reputation/Coreputation.sol:26]: Done!

**Review by benjaminbollen — APPROVED** [review]
LGTM

**Pull request merged**